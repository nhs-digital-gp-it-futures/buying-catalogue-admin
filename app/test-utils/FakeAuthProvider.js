export class FakeAuthProvider {
  // eslint-disable-next-line no-useless-constructor, no-empty-function
  constructor(fakeLogoutMethod = () => {}) {
    this.fakeLogout = fakeLogoutMethod;
  }

  // eslint-disable-next-line class-methods-use-this
  setup(app) {
    app.use((req, res, next) => {
      if (req.cookies && req.cookies.fakeToken) {
        req.user = JSON.parse(req.cookies.fakeToken);
      }
      req.logout = this.fakeLogout;
      next();
    });
  }

  // eslint-disable-next-line class-methods-use-this, no-unused-vars
  login() {
    return (req, res) => {
      res.redirect('http://identity-server/login');
    };
  }

  // eslint-disable-next-line class-methods-use-this, no-unused-vars
  loginCallback() {
    return (req, res) => {
      res.redirect('/');
    };
  }

  // eslint-disable-next-line class-methods-use-this
  logout() {
    return '/signout-callback-oidc';
  }

  // eslint-disable-next-line class-methods-use-this
  authorise() {
    return (req, res, next) => {
      if (!req.user) {
        this.login()(req, res, next);
      } else if (req.user && req.user.organisation) {
        next();
      } else {
        throw new Error('Not authorised matey');
      }
    };
  }
}
