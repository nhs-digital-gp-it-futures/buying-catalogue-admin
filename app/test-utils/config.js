export const apiLocalhost = 'http://localhost:5100';

export const organisationsApiLocalhost = 'http://localhost:8075';

export const apiPath = '/api/v1';

export const clientLocalhost = 'http://localhost:1234';
