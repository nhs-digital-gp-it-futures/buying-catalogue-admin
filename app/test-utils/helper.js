import cheerio from 'cheerio';
import request from 'supertest';

export const extractInnerText = async (element) => {
  const elementInnerText = await element.innerText;
  return elementInnerText.trim();
};

const extractCsrfToken = (res) => {
  const $ = cheerio.load(res.text);
  return $('[name=_csrf]').val();
};

export const getCsrfTokenFromGet = async (app, getPath, authenticationCookie) => {
  let cookies;
  let csrfToken;

  await request(app)
    .get(getPath)
    .set('Cookie', authenticationCookie)
    .then((getRes) => {
      cookies = getRes.headers['set-cookie'];
      csrfToken = extractCsrfToken(getRes);
    });
  return { cookies, csrfToken };
};

export const setFakeCookie = async (app, getPath) => {
  let cookies;
  await request(app)
    .get(getPath)
    .set('Cookie', ['cookie1=cookie1value'])
    .then((getRes) => {
      cookies = getRes.headers['set-cookie'];
    });

  return { modifiedApp: app, cookies };
};
