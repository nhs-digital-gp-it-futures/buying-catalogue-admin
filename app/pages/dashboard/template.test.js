import { createTestHarness } from '../../test-utils/testHarness';
import context from './manifest.json';

context.backLinkHref = '/';

const setup = {
  template: {
    path: 'pages/dashboard/template.njk',
  },
};

describe('organisations dashboard page', () => {
  it('should render a backLink to the home page', createTestHarness(setup, (harness) => {
    harness.request(context, ($) => {
      const homepageBackLink = $('[data-test-id="go-back-link"]');
      expect(homepageBackLink.length).toEqual(1);
      expect(homepageBackLink.text().trim()).toEqual('Go back to previous page');
      expect($(homepageBackLink).find('a').attr('href')).toEqual('/');
    });
  }));

  it('should render a title', createTestHarness(setup, (harness) => {
    harness.request(context, ($) => {
      const title = $('h1[data-test-id="org-dashboard-title"]');
      expect(title.length).toEqual(1);
      expect(title.text().trim()).toEqual(context.title);
    });
  }));

  it('should render a description', createTestHarness(setup, (harness) => {
    harness.request(context, ($) => {
      const description = $('[data-test-id="org-dashboard-description"]');
      expect(description.length).toEqual(1);
      expect(description.text().trim()).toEqual(context.description);
    });
  }));

  it('should render a add organisation button', createTestHarness(setup, (harness) => {
    harness.request(context, ($) => {
      const addOrgButton = $('[data-test-id="add-org-button"] a');
      expect(addOrgButton.length).toEqual(1);
      expect(addOrgButton.text().trim()).toEqual(context.addOrgButtonText);
      expect(addOrgButton.attr('href')).toEqual('#');
    });
  }));

  it('should render the table component', createTestHarness(setup, (harness) => {
    harness.request(context, ($) => {
      const orgTable = $('[data-test-id="org-table"]');
      expect(orgTable.length).toEqual(1);
      expect(orgTable.find('[data-test-id="table"]').length).toEqual(1);
      expect(orgTable.find('[data-test-id="table-headings"]').length).toEqual(1);
      expect(orgTable.find('[data-test-id="column-heading"]').length).toEqual(2);
      // TODO: Add tests to check the organisations data comes through
    });
  }));
});
