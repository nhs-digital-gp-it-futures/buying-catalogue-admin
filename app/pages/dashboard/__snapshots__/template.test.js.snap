// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`organisations dashboard page the page should render 1`] = `
"<main class=\\"nhsuk-main-wrapper\\" id=\\"maincontent\\" data-test-id=\\"main-content\\">
        
  <div class=\\"nhsuk-width-container\\" data-test-id=\\"organisations\\">
    <div data-test-id=\\"go-back-link\\">
      <div class=\\"nhsuk-back-link\\">
  <a class=\\"nhsuk-back-link__link\\" href=\\"http://localhost:3000/re-login\\">
  <svg class=\\"nhsuk-icon nhsuk-icon__chevron-left\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" height=\\"24\\" width=\\"24\\">
    <path d=\\"M8.5 12c0-.3.1-.5.3-.7l5-5c.4-.4 1-.4 1.4 0s.4 1 0 1.4L10.9 12l4.3 4.3c.4.4.4 1 0 1.4s-1 .4-1.4 0l-5-5c-.2-.2-.3-.4-.3-.7z\\"></path>
  </svg>
  Go back to previous page</a>
</div>

    </div>
    <h1 data-test-id=\\"org-dashboard-title\\" class=\\"nhsuk-u-margin-top-7 nhsuk-u-margin-bottom-3\\">Manage organisations and users</h1>
    <p data-test-id=\\"org-dashboard-description\\" class=\\"nhsuk-heading-l nhsuk-u-font-size-22 nhsuk-u-margin-bottom-9\\">Edit existing account information or add new users to an organisation.</p>

    <div data-test-id=\\"add-org-button\\">
      


  
    
  




<a class=\\"nhsuk-button nhsuk-u-margin-bottom-9\\" href=\\"/organisations/find\\" draggable=\\"false\\">
  Add an organisation
</a>



    </div>

    <div data-test-id=\\"org-table\\" class=\\"nhsuk-u-font-size-19\\">
      
  <table data-test-id=\\"table\\" class=\\"\\" style=\\"page-break-inside:auto\\">
    
    
      <thead data-test-id=\\"table-headings\\" style=\\"display:table-header-group;\\">
        
          <tr><th data-test-id=\\"column-heading-0\\" class=\\"nhsuk-u-padding-bottom-0 \\" style=\\"width:\\">
            <div data-test-id=\\"column-heading-0-data\\" class=\\"nhsuk-u-font-weight-bold\\">Organisation name</div>
            
            
          </th>
        
          <th data-test-id=\\"column-heading-1\\" class=\\"nhsuk-u-padding-bottom-0 \\" style=\\"width:\\">
            <div data-test-id=\\"column-heading-1-data\\" class=\\"nhsuk-u-font-weight-bold\\">ODS code</div>
            
            
          </th>
        
      </tr></thead>
    

    <tbody>
    
      <tr data-test-id=\\"table-row-0\\" style=\\"page-break-inside:avoid; page-break-after:auto\\">
        
          
          <td class=\\"nhsuk-u-padding-bottom-0\\" style=\\"border-bottom-style: solid\\">

            
              <a data-test-id=\\"org-name-org1\\" href=\\"/organisations/org1\\">A lovely organisation</a>
            

            
          </td>
        
          
          <td class=\\"nhsuk-u-padding-bottom-0\\" style=\\"border-bottom-style: solid\\">

            
              
  
    
    <div data-test-id=\\"ods-code-org1\\" class=\\"bc-c-data-text \\">ODS1</div>
  

            

            
          </td>
        
      </tr>
    
      <tr data-test-id=\\"table-row-1\\" style=\\"page-break-inside:avoid; page-break-after:auto\\">
        
          
          <td class=\\"nhsuk-u-padding-bottom-0\\" style=\\"border-bottom-style: solid\\">

            
              <a data-test-id=\\"org-name-org2\\" href=\\"/organisations/org2\\">Another lovely organisation</a>
            

            
          </td>
        
          
          <td class=\\"nhsuk-u-padding-bottom-0\\" style=\\"border-bottom-style: solid\\">

            
              
  
    
    <div data-test-id=\\"ods-code-org2\\" class=\\"bc-c-data-text \\">ODS2</div>
  

            

            
          </td>
        
      </tr>
    
    </tbody>
    
  </table>

    </div>
  </div>

      </main>"
`;
